ansible/roles/ci/code_coverage/xml/files/ament_xml_codecov.py

```diff
@@ -0,0 +1,301 @@
+
+        dependencies = gather_all_dependencies(filename)
+        if dependencies:
+            errors = test_dependencies(dependencies, filename)
+        else:
+            errors = None
+
+        filename = os.path.relpath(filename, start=os.getcwd())
+        report.append((filename, errors))
+    if args.debug:
+        for (filename, errors) in report:
+            if errors is not None:
+                errormsg = "File '{}' is invalid: ".format(filename)
+                for line in errors.splitlines():
+                    errormsg += line.strip() + " "
+                print(errormsg, file=sys.stderr)
+                print('', file=sys.stderr)
+        # output summary
+    error_count = sum(1 if r[1] else 0 for r in report)
+    if not error_count:
+        if args.debug:
+            print('No problems found')
+        rc = 0
+    else:
+        if args.debug:
+            print('%d files are invalid' % error_count, file=sys.stderr)
+        rc = 1
+    # generate xunit file
+    if args.xunit_file:
+        folder_name = os.path.basename(os.path.dirname(args.xunit_file))
+        file_name = os.path.basename(args.xunit_file)
+        suffix = '.xml'
+        if file_name.endswith(suffix):
+            file_name = file_name[0:-len(suffix)]
+            suffix = '.xunit'
+            if file_name.endswith(suffix):
+                file_name = file_name[0:-len(suffix)]
+        testname = '%s.%s' % (folder_name, file_name)
+
+        xml = get_xunit_content(report, testname, time.time() - start_time)
+        path = os.path.dirname(os.path.abspath(args.xunit_file))
+        if not os.path.exists(path):
+            os.makedirs(path)
+        with open(args.xunit_file, 'w') as f:
+            f.write(xml)
+    return rc
+
+
+def gather_files(directory, extensions):
+    """Walks through the directory and puts all files with the correct
+    extensions into a list, exits if empty."""
+    all_files = []
+    for root, _, files in os.walk(directory):
+        for file in files:
+            if os.path.splitext(file)[1][1:] in extensions:
+                all_files.append(os.path.join(root, file))
+    if not all_files:
+        print("No files detected.\nExiting test.")
+        exit(0)
+    return all_files
+
+
+def gather_all_dependencies(filename):
+    """Goes through each file and gathers all includes within the file."""
+    try:
+        tree = ElementTree.parse(filename)
+    except ElementTree.ParseError:  # If file doesn't parse it's caught by previous check.
+        return None
+    root = tree.getroot()
+    dependencies = []
+    for dep in root.findall('include'):
+        fulldep = ElementTree.tostring(dep).decode("utf-8").split('\n')[0]
+        dependencies.append((fulldep, dep.attrib['file']))
+    for dep in root.findall('xacro:include'):
+        fulldep = ElementTree.tostring(dep).decode("utf-8").split('\n')[0]
+        dependencies.append((fulldep, dep.attrib['filename']))
+    return dependencies
```

Comment:
same as below

Reply:
Do you mean the fact its replicated code? Because they are looking for different elements within the xml

Reply:
thought about the try: except: but you proved me wrong in previous comment. All fine now :)